leakcanary-hprof/src/main/java/leakcanary/hprof/HProfParser.java

```diff
@@ -0,0 +1,1109 @@
+          readId(hprofSource); // protection domain object ID
+          readId(hprofSource); // reserved
+          readId(hprofSource); // reserved
+          int instanceSize = hprofSource.readInt();
+          numBytes -= 7 * idSize + 8;
+
+          log(3, "tag", "CLASS DUMP");
+          log(4, "class", getClassName(classObjectId));
+          log(4, "super class", getClassName(superClassObjectID));
+          // log(4, "class loader", getStringName(classLoaderObjectID));
+          log(4, "instance size", instanceSize);
+
+          short constantPoolSize = hprofSource.readShort();
+          numBytes -= 2;
+
+          if (constantPoolSize != 0) {
+            log(4, "constants", "");
+
+            for (int i = 0; i < constantPoolSize; i++) {
+              short constantPoolIndex = hprofSource.readShort();
+              Type type = Type.getType(hprofSource.readByte());
+              Object value = readTypeElement(hprofSource, type);
+              numBytes -= 3 + getTypeSize(type);
+              log(5, "#" + constantPoolIndex, type + " -> " + value);
+            }
+          }
+
+          short numStaticFields = hprofSource.readShort();
+          numBytes -= 2;
+
+          if (numStaticFields != 0) {
+            log(4, "static fields", "");
+
+            for (int i = 0; i < numStaticFields; i++) {
+              long staticFieldNameStringID = readId(hprofSource);
+              Type type = Type.getType(hprofSource.readByte());
+              Object value = readTypeElement(hprofSource, type);
+              numBytes -= idSize + 1 + getTypeSize(type);
+              log(5, type + " " + stringMap.get(staticFieldNameStringID), value);
+            }
+          }
+
+          short numInstanceFields = hprofSource.readShort();
+          numBytes -= 2;
+
+          InstanceField[] instanceFields = new InstanceField[numInstanceFields];
+          if (numInstanceFields != 0) {
+            log(4, "instance fields", "");
+            for (int i = 0; i < numInstanceFields; i++) {
+              long fieldNameStringID = readId(hprofSource);
+              Type type = Type.getType(hprofSource.readByte());
+              instanceFields[i] = new InstanceField(fieldNameStringID, type);
+              numBytes -= idSize + 1;
+              log(5, type + " " + stringMap.get(fieldNameStringID), "");
+            }
+          }
+          ClassInfo classInfo = new ClassInfo(classObjectId, superClassObjectID, instanceSize, instanceFields);
+          classInfoMap.put(classObjectId, classInfo);
+
+          break;
+        }
+        case INSTANCE_DUMP: {
+          INSTANCE_COUNT++;
+
+          long objectId = readId(hprofSource);
+          hprofSource.readInt(); // stack trace serial #
+          long classObjectId = readId(hprofSource);
+          int numInstanceBytes = hprofSource.readInt();
+
+          // TODO - process instance values afterward
+          byte[] instanceFieldValueBytes = hprofSource.readByteArray(numInstanceBytes);
```

Comment:
`readByteString` will likely let you share the backing `byte[]` so you're not making a copy here